File: client-clojure/src/sirqul_io_t_platform/specs/game.clj
(ns sirqul-io-t-platform.specs.game
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.application :refer :all]
            [sirqul-io-t-platform.specs.pack :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.game-level :refer :all]
            )
  (:import (java.io File)))


(def game-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :owner) account-spec
   (ds/opt :application) application-spec
   (ds/opt :title) string?
   (ds/opt :description) string?
   (ds/opt :startDate) inst?
   (ds/opt :endDate) inst?
   (ds/opt :packs) (s/coll-of pack-spec)
   (ds/opt :sequenceType) string?
   (ds/opt :authorOverride) string?
   (ds/opt :icon) asset-spec
   (ds/opt :image) asset-spec
   (ds/opt :points) int?
   (ds/opt :ticketType) string?
   (ds/opt :ticketCount) int?
   (ds/opt :allocateTickets) boolean?
   (ds/opt :allGameLevels) (s/coll-of game-level-spec)
   (ds/opt :contentName) string?
   (ds/opt :contentType) string?
   })

(def game-spec
  (ds/spec
    {:name ::game
     :spec game-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/Game.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Game._

case class Game (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
owner: Option[Account],
application: Option[Application],
title: Option[String],
description: Option[String],
startDate: Option[OffsetDateTime],
endDate: Option[OffsetDateTime],
packs: Option[List[Pack]],
sequenceType: Option[SequenceType],
authorOverride: Option[String],
icon: Option[Asset],
image: Option[Asset],
points: Option[Long],
ticketType: Option[String],
ticketCount: Option[Long],
allocateTickets: Option[Boolean],
allGameLevels: Option[List[GameLevel]],
contentName: Option[String],
`contentType`: Option[String])

object Game {
  import DateTimeCodecs._
  sealed trait SequenceType
  case object FIRSTAVAILABLE extends SequenceType
  case object ALLAVAILABLE extends SequenceType

  object SequenceType {
    def toSequenceType(s: String): Option[SequenceType] = s match {
      case "FIRSTAVAILABLE" => Some(FIRSTAVAILABLE)
      case "ALLAVAILABLE" => Some(ALLAVAILABLE)
      case _ => None
    }

    def fromSequenceType(x: SequenceType): String = x match {
      case FIRSTAVAILABLE => "FIRSTAVAILABLE"
      case ALLAVAILABLE => "ALLAVAILABLE"
    }
  }

  implicit val SequenceTypeEnumEncoder: EncodeJson[SequenceType] =
    EncodeJson[SequenceType](is => StringEncodeJson(SequenceType.fromSequenceType(is)))

  implicit val SequenceTypeEnumDecoder: DecodeJson[SequenceType] =
    DecodeJson.optionDecoder[SequenceType](n => n.string.flatMap(jStr => SequenceType.toSequenceType(jStr)), "SequenceType failed to de-serialize")

  implicit val GameCodecJson: CodecJson[Game] = CodecJson.derive[Game]
  implicit val GameDecoder: EntityDecoder[Game] = jsonOf[Game]
  implicit val GameEncoder: EntityEncoder[Game] = jsonEncoderOf[Game]
}
